Support key binding for Bash 3

akinomyoga opened the pull request:
Hi, @urbainvaes @3ximus! I was playing with the latest version. This is a further improvement for #41 and #43. I think I managed to support working key bindings for Bash 3. This PR also includes miscellaneous other fixes. These fixes are not independent of one another, so I gave up to separate the PR into smaller ones. Here are the descriptions for each fix:

<details><summary>[ <b>Outdated</b>: I have rebased the commits. See <a href="#user-content-current-patch-set">current patch set</a>. ]</summary>

- Commit 4deba7c
  - It turned out that Bash 3 and 4 don't work with two-byte intermediate keys such as `mark1='\302\200'` which is used when the character encoding of the current locale is UTF-8. For UTF-8, I switched to use the bytes that will never be used in UTF-8, such as `\xC0`, `\xC1`, `\xF5`-`\xFF`. In addition, it seems that Bash 4 will unexpectedly terminate when it receives `\xFE` or `\xFF`, so finally, I decided to use `\xC0`, `\xC1` and `\xFD`.
  - For use in the later commits, I made the variables `mark{1..3}` global after renaming them to `_fzm_key{1..3}`.
- Commit 9f771d2
  - I added a support for Bash 3 so that the original contents and the cursor position of the command line is preserved after the invocation of `fzm`. It uses the same technique as `fzf` uses for Bash 3, which assumes that the key bindings to <kbd>C-u</kbd>, <kbd>C-e</kbd>, <kbd>C-k</kbd>, <kbd>C-y</kbd>, <kbd>M-y</kbd>, <kbd>M-space</kbd>, <kbd>C-x C-x</kbd>, and <kbd>C-m</kbd> are the Bash default.
  - For Bash 4, I added a workaround for the bashbug of `READLINE_POINT` (c.f. [bug-bash Mailing list - 2018-04/40](https://lists.gnu.org/archive/html/bug-bash/2018-04/msg00040.html)).
  - For Bash 5+, I added support for the variable `READLINE_MARK`, the new feature of Bash 5.0.
  - For Bash 4+ and `ble.sh` bindings, it is changed to go to the next line only when the current working directory is changed.
- Commit c0f299b
  - Now, the paste key (`ctrl-k` by default) is only enabled when `fzm` is used from key bindings, i.e., the paste key is disabled when `fzm` is executed from the command line.
  - The description of the key configuration displayed in the header of `fzf` is now properly set depending on the context (such as `fzf`, `jump`, `dmark`, or key bindings).
- Commit d747154
  -  `FZF_MARKS_JUMP` is now used also for the `ble.sh` binding. To support this, I updated `ble.sh` with the commit https://github.com/akinomyoga/ble.sh/commit/0f01cab8dce267fe75fa92c082c28b0d05c3d885 so that `ble-bind` accepts the Readline format (i.e., `bind` format) of key sequences with the type specifier `keyseq:`. Thus, this needs to be used with the latest devel version (`master` branch) of `ble.sh`.
- Commit 9f27330
  - Use `bind -x` for `FZF_MARKS_DMARK`
 </details>

Sorry for the relatively large set of changes. You can specify which changes from the above should be picked and which changes should be rejected so that I can adjust the related commits. Also, any comments, suggestions and questions are welcome!

Thank you

----

## Current Patch Set (Update 2021-01-03 04:54:31 JST)<a id="current-patch-set">†</a>

- Commit 4deba7c
  - It turned out that Bash 3 and 4 don't work with two-byte intermediate keys such as `mark1='\302\200'` which is used when the character encoding of the current locale is UTF-8. For UTF-8, I switched to use the bytes that will never be used in UTF-8, such as `\xC0`, `\xC1`, `\xF5`-`\xFF`. In addition, it seems that Bash 4 will unexpectedly terminate when it receives `\xFE` or `\xFF`, so finally, I decided to use `\xC0`, `\xC1` and `\xFD`.
  - For use in the later commits, I made the variables `mark{1..3}` global after renaming them to `_fzm_key{1..3}`.
- Commit 1d88f86
  - I added a support for Bash 3 so that the original contents and the cursor position of the command line is preserved after the invocation of `fzm`. It uses the same technique as `fzf` uses for Bash 3, which assumes that the key bindings to <kbd>C-u</kbd>, <kbd>C-e</kbd>, <kbd>C-k</kbd>, <kbd>C-y</kbd>, <kbd>M-y</kbd>, <kbd>M-space</kbd>, <kbd>C-x C-x</kbd>, <kbd>C-q</kbd>, <kbd>C-?</kbd>, <kbd>C-d</kbd>, and <kbd>C-m</kbd> are the Bash default.
  - For Bash 4, I added a workaround for the bashbug of `READLINE_POINT` (c.f. [bug-bash Mailing list - 2018-04/40](https://lists.gnu.org/archive/html/bug-bash/2018-04/msg00040.html)).
  - For Bash 5+, I added support for the variable `READLINE_MARK`, the new feature of Bash 5.0.
  - For Bash 4+ and `ble.sh` bindings, it is changed to go to the next line only when the current working directory is changed.
  - Now, `fzm` prints the selected path for `ctrl-k` by default when used outside the key bindings.
  - The command to be used for the "paste" operation outside the key bindings can be now configured with the variable `FZF_MARKS_PASTE_COMMAND`. When the variable is empty, `printf '%s\n'` will be used.
- Commit 0405ea1
  -  `FZF_MARKS_JUMP` is now used also for the `ble.sh` binding. To support this, I updated `ble.sh` with the commit https://github.com/akinomyoga/ble.sh/commit/0f01cab8dce267fe75fa92c082c28b0d05c3d885 so that `ble-bind` accepts the Readline format (i.e., `bind` format) of key sequences with the type specifier `keyseq:`. Thus, this needs to be used with the latest devel version (`master` branch) of `ble.sh`.
- Commit 873d7ba
  - Use `bind -x` for `FZF_MARKS_DMARK`
- Commit e844e30
  - Fix the vulnerability against code injection. For example, `jump '$(echo hello)'` causes an execution of `$(echo hello)` before the startup of `fzf`. For another example, `jump ' | echo hello '` would cause strange behavior.
- Commit 07d498e
  - I've escalated `pmark` as an independent command (just like `jump` and `dmark`) to print/paste the selected directory.
- Commit d15e285
  - The description of the key configuration displayed in the header of `fzf` is now properly set depending on the context (such as `fzf`, `jump`, `dmark`, or key bindings).
- Commit 6abe218
  - I changed `pmark` to output to stdout by default when the stdout is not TTY.
- Commit 95aa426
  - The default paste key `ctrl-k` has been changed to `ctrl-v`.

3ximus commented:
Hi there!
Yeah, I tested on bash 4 now and indeed the binding wasn't working. Everything works fine now. :+1: 

Why did you decide to remove the `ctrl+k` binding  though? Personally I'll never use this without the bindings, but it removes the ability to do `ls $(fzm)` if someone wants to do it...

akinomyoga commented:
> Why did you decide to remove the ctrl+k binding though? Personally I'll never use this without the bindings, but it removes the ability to do ls $(fzm) if someone wants to do it...

Oh, I see! I didn't understand the intent of printing the selected path! I'll fix it to output the path when used as `$(fzm)`, etc.

akinomyoga commented:
Thank you! I added the following changes:

- Commit bc6d71b
  - Now, `fzm` prints the selected path by default when used outside the key bindings.
  - The command to be used for the "paste" operation outside the key bindings can be now configured with the variable `FZF_MARKS_PASTE_COMMAND`. When the variable is empty, `printf '%s\n'` will be used.
- Commit f04e1ba
  - I fixed the problem in Bash 3 that old commandline contents are restored when the cursor is in the beginning or the end of the line and the current working directory is changed by the `fzm` binding.

By the way, when `fzm` is used as `$(fzm)` or `fzm | command` (i.e., when stdout is not connected to TTY), maybe `fzm` should output the path by default instead of performing `jump`. What do you think?

3ximus commented:
Yeah, i think it would make more sense. But maybe still keep the option to make the behavior consistent?

akinomyoga commented:
> But maybe still keep the option to make the behavior consistent?

Yeah, that seems better. What kind of an option would you suggest? For example, some variable like  `FZF_MARKS_FZM_DEFAULT=always-jump`?

3ximus commented:
I just meant the option to keep using the `ctrl+k` binding, I agree with the default paste from the terminal :+1: 
But I don't really see many uses for this feature anyway so its up to you xD

akinomyoga commented:
> I just meant the option to keep using the ctrl+k binding,

I see. In that point, I agree with you to keep `ctrl-k` even with `$(fzm)` and `fzm | command` (which means `ctrl-m`, `ctrl-y`, and `ctrl-k` behaves the same in `$(fzm)` and `fzm | command`). In commit 9edc1c9, I changed to output to stdout by default when the stdout is not TTY. These are the additional fixes:

- Commit 706d093
  - I applied the same change to zsh config as that of bash config in c0f299b
- Commit 1a9e859
  - Fix the vulnerability against code injection. For example, `jump '$(echo hello)'` causes an execution of `$(echo hello)` before the startup of `fzf`. For another example, `jump ' | echo hello '` would cause strange behavior.
- Commit 208163a
  - I've escalated `pmark` as an independent command (just like `jump` and `dmark`) to print/paste the selected directory, and added the function also to zsh config.

Maybe I'll rebase and squash closely-related commits later.

urbainvaes commented:
Hi @akinomyoga and @3ximus!Thanks for all the good work!

I was thinking that it might be good to use something other than `ctrl-k` for pasting, because some people might be using `ctrl-j` or `ctrl-k` for going down and up the list (personally I prefer `ctrl-n` and `ctrl-p`). What do you think?

akinomyoga commented:
OK, `ctrl-k` is bounded by going up by default in `fzf`, so it seems better to keep its original meaning in the default settings.
- Vim uses `p` and `P` for pasting which would correspond to the key `ctrl-p` in `fzf`, but `ctrl-p` is already used for going up in `fzf`.
- Emacs uses `ctrl-y` for paste, but `ctrl-y` has been already the accept key of `fzf-marks` for a long time.
- Windows applications typically use `ctrl-v` for paste.
- X11 applications typically use `shift-insert` for paste. Maybe `insert` can also be used. The problem is that not all the keyboard has the `insert` key.

`ctrl-v`?

urbainvaes commented:
`ctrl-v` seems good to me!

3ximus commented:
Oh wow, when I choose `ctrl+k` i totally forgot you could use them to navigate and i've been using `ctrl+n` and `ctrl+p`. But I agree, the default should change to `ctrl+v`.

akinomyoga commented:
Thank you! I changed the default paste key to `ctrl-v` in commit 95aa426. I also rebased and squashed commits and then force-pushed. The descriptions on the current patch set is summarized in [the list in the top comment](#user-content-current-patch-set).

akinomyoga commented on fzf-marks.plugin.bash:
```diff
@@ -36,7 +36,7 @@ if [[ -z "${FZF_MARKS_COMMAND}" ]] ; then
 
     if [[ $FZF_VERSION -gt $MINIMUM_VERSION ]]; then
-        FZF_MARKS_COMMAND="fzf --height 40% --reverse --header='ctrl-y:jump, ctrl-t:toggle, ctrl-d:delete, ctrl-k:paste'"
+        FZF_MARKS_COMMAND="fzf --height 40% --reverse --header=\"\$_fzm_keymap_description\""
```
One thing that I want to ask is whether there is a reason to include the `--header` option in `FZF_MARKS_COMMAND`. Why don't we specify the `--header` option in the caller side?

urbainvaes replied:
Could you elaborate? I am not sure that I understand the question...

akinomyoga replied:
Sorry for my unclear question. Currently, the option `--header="$_fzm_keymap_description"` is specified in `FZF_MARKS_COMMAND`, and the actual contents of the header are specified at the caller sides (`fzm`, `jump`, `dmark`, and `pmark`), whose structure can be illustrated by the following code:

```bash
# Initialize
FZF_MARKS_COMMAND='fzf --height 40% --reverse --header="$_fzm_keymap_description"'

# Caller side
local _fzm_keymap_description='<HEADER_CONTENT>'
eval "$FZF_MARKS_COMMAND" --more-options...
```

I thought it is cleaner to directly specify the header option at the caller side in the following way:

```bash
# Initialize
FZF_MARKS_COMMAND='fzf --height 40% --reverse'

# Caller side
eval "$FZF_MARKS_COMMAND" --header='<HEADER_CONTENT>' --more-options...
```

But after I asked the question, I checked the minimum version for the `--header` option, and it was `0.10.5` (https://github.com/junegunn/fzf/commit/2022a3a on 2015-09-15) which is later than the project start of `fzf-marks` (initial commit 2015-06-04). Maybe this is the reason that `--header` is included in `FZF_MARKS_COMMAND`? (Another option included in `FZF_MARKS_COMMAND`, `--height` which is supported from `0.16.00` (https://github.com/junegunn/fzf/commit/1448d631 2017-01-08), is used only when `[[ $FZF_VERSION -gt $MINIMUM_VERSION ]]`.)

Thank you

urbainvaes replied:
Hi @akinomyoga ,

I understand what you mean now, and I agree that your proposed solution seems cleaner. In the pull request that introduced `--header`, the header was hardcoded, so I don't think there was a version issue.

akinomyoga replied:
Oh, I see! Thank you for pointing to the background of introducing `--header`! Then, I'll make another PR for this refactoring. Besides this one, I also have two other small fixes for which I'll open PRs soon. Thank you again!

akinomyoga commented on fzf-marks.plugin.bash:
```diff
@@ -171,18 +171,26 @@ function set-up-fzm-bindings {
+        # Intiialize special keys used for key bindings
+        _fzm_key1='\200'
+        _fzm_key2='\201'
+        _fzm_key3='\202'
```
Here I chose `_fzm_*` as the prefix for the private global variables used by `fzf-marks`. If you have any suggestion on the naming convention on the private variables, please let me know.

urbainvaes replied:
Seems clean to me!

akinomyoga commented on fzf-marks.plugin.bash:
```diff
@@ -165,11 +159,148 @@ function dmark {
+    # Widget for Bash 4.0+
+
+    # bashbug https://lists.gnu.org/archive/html/bug-bash/2018-04/msg00040.html
+    function _fzm-widget-has_readline_point_bug {
```
Also, do you have any suggestions for the name of the private `fzm-marks` functions? Currently, I use `_fzm` as the private prefix and connect names with hyphens. However, the conventional naming of shell functions only uses the underscore in addition to the alphanumeric letters but not the hyphen.

urbainvaes replied:
I think `_fzm` is very good!

akinomyoga replied:
Thank you! I also changed the names of the other private functions (the functions of internal use of `fzf-marks`). I think we should add the prefix to all the internal functions. For example, `setup_complete` could conflict with the function name from the other plugins or user settings of a similar purpose.

urbainvaes replied:
Now that you point this out, I completely agree! Thank you!

akinomyoga reviewed (commented):
@urbainvaes I have several questions on the design and coding styles as follows. If you have any thoughts, please let me know!

urbainvaes commented:
@3ximus, @akinomyoga: Thank you very much again for all your excellent work! The new features work very well thanks to you, and you both have a great knowledge of shell scripting, so I'm really glad that you contributed to this project! :)

I am merging this already, but feel free to reopen another pull request if you'd like to propose other changes.
